gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/Misc.java

```diff
@@ -1372,4 +1374,38 @@ public static void invalidSnappyDataFeature(String featureDescription)
               "This feature is supported when product is started in rowstore mode");
     }
   }
+
+  public static void checkClusterRestrictions(GemFireStore.VMKind vmKind) {
+    if (/*Misc.getMemStore().isSnappyStore() &&*/ GemFireVersion.isEnterpriseEdition() &&
+        vmKind == GemFireStore.VMKind.DATASTORE && isAWS() && isRestrictedCluster()) {
+      int servers = 0;
+      Set<DistributedMember> all = Misc.getDistributedSystem().getAllOtherMembers();
+      for (DistributedMember dm : all) {
+        if (GemFireXDUtils.getVMKind(dm) == GemFireStore.VMKind.DATASTORE) {
+          servers++;
+        }
+      }
+      if (servers > 1) { // Only allow max 2 dataservers (or one?)
+        throw new IllegalStateException("Dataserver cannot be started as the dataserver limit " +
+            "exceeds.");
+      }
+    }
+  }
+
+  private static boolean isRestrictedCluster() {
+    return "singleNode".equalsIgnoreCase(com.pivotal.gemfirexd.internal.snappy.CallbackFactoryProvider.getClusterCallbacks()
+        .getClusterType());
+  }
+
+  private static boolean isAWS() {
+    try {
+      String cmd = "curl --connect-timeout 2 http://169.254.169.254/latest/meta-data/ami-id";
+      Process p = Runtime.getRuntime().exec(cmd);
+      p.waitFor(3, TimeUnit.SECONDS);
+      return p.exitValue() == 0;
```

Comment:
Is just the exitValue enough or need to check output too? The wait might be increased a bit. Also since this is an expensive method, it should be done only once in static block or something.